### PR TITLE
modern cpp solution

### PR DESCRIPTION
**Pull request by marcindus**

### CODE REVIEW

File: shapes/Circle.cpp [hunk 1]
```diff
@@ -6,6 +6,13 @@ Circle::Circle(double r)
     : r_(r)
 {}
 
+
+
+Circle::Circle(double r, Color c = Color::Green) : r_(r), Shape(c)
+{
+
```
Comment (MaciejSalamonski) [inline shapes/Circle.cpp]: To mamy empty lines there

File: shapes/main.cpp [hunk 2]
```diff
@@ -63,18 +63,22 @@ void findFirstShapeMatchingPredicate(const Collection& collection,
     }
 }
 
-constexpr int fibo(int n) {
-    if (n<=2) {
+constexpr int fibo(int n)
+{
+    if (n <= 2)
```
Comment (MaciejSalamonski) [inline shapes/main.cpp]: Maybe let's use operator "?"

**Review by MaciejSalamonski — APPROVED** [review]

**Review by MaciejSalamonski — APPROVED** [review]
:+1:

File: shapes/main.cpp [hunk 3]
```diff
@@ -1,58 +1,58 @@
...
+template<class DerivedType, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args);
+
 
 void printCollection(const Collection& collection)
 {
-    for (const auto & it : collection)
+    for (const auto& it : collection)
         if (it)
             it->print();
```
Comment (WBigaj) [inline shapes/main.cpp:35]:
```suggestion
    for (const auto& it : collection) {
        if (it) {
            it->print();
        }
    }
```

File: shapes/main.cpp [hunk 4]
```diff
@@ -1,58 +1,58 @@
...
 }
 
 void printAreas(const Collection& collection)
 {
-    for (const auto & it : collection)
+    for (const auto& it : collection)
         if (it)
             cout << it->getArea() << std::endl;
```
Comment (WBigaj) [inline shapes/main.cpp:42]:
```suggestion
    for (const auto& it : collection) {
        if (it) {
            cout << it->getArea() << std::endl;
        }
    }
```

File: shapes/main.cpp [hunk 5]
```diff
@@ -63,18 +63,22 @@ void findFirstShapeMatchingPredicate(const Collection& collection,
     }
 }
 
-constexpr int fibo(int n) {
-    if (n<=2) {
+constexpr int fibo(int n)
+{
+    if (n <= 2)
+    {
         return 1;
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
```

File: shapes/main.cpp [hunk 6]
```diff
@@ -63,18 +63,22 @@ void findFirstShapeMatchingPredicate(const Collection& collection,
...
-    } else {
+    }
+    else
+    {
         return fibo(n - 1) + fibo(n - 2);
     }
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    }
    else
    {
        return fibo(n - 1) + fibo(n - 2);
    }
    return (n <= 2) ? 1 : (fibo(n - 1) + fibo(n - 2));
```

Comment (WBigaj) [reply shapes/main.cpp]: Speaking of @MaciejSalamonski advice.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: shapes/main.cpp [hunk 7]
```diff
@@ -1,58 +1,58 @@
...
 }
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
-    return false;
-}
+template<class DerivedType, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args);
```
Comment (ziobron) [inline shapes/main.cpp]: forward declaration below actual function definition. To be deleted.

**Review by ziobron — CHANGES_REQUESTED** [review]
Everything is ok, but you have not used delegating constructors. This feature allow you to call other class constructor on constructor initalizer list. It wasn't allowed before C++14. You could only call Base class constructor.

**Comment by marcindus** [comment]
> Everything is ok, but you have not used delegating constructors. This feature allow you to call other class constructor on constructor initalizer list. It wasn't allowed before C++14. You could only call Base class constructor.

 I've added deleating constructor in Circle class: Base class shape constructor call:  Circle::Circle(double r, Color c = Color::Green) : r_(r), Shape(c)

**Comment by ziobron** [comment]
> > Everything is ok, but you have not used delegating constructors. This feature allow you to call other class constructor on constructor initalizer list. It wasn't allowed before C++14. You could only call Base class constructor.
> 
> I've added deleating constructor in Circle class: Base class shape constructor call: Circle::Circle(double r, Color c = Color::Green) : r_(r), Shape(c)

It is not delegating constructor. It is calling a base class constructor. Constructor delegation is when you call constructor of the same class on the initalizer list of other constructor

https://www.geeksforgeeks.org/constructor-delegation-c/

**Comment by marcindus** [comment]
> > > Everything is ok, but you have not used delegating constructors. This feature allow you to call other class constructor on constructor initalizer list. It wasn't allowed before C++14. You could only call Base class constructor.
> > 
> > 
> > I've added deleating constructor in Circle class: Base class shape constructor call: Circle::Circle(double r, Color c = Color::Green) : r_(r), Shape(c)
> 
> It is not delegating constructor. It is calling a base class constructor. Constructor delegation is when you call constructor of the same class on the initalizer list of other constructor
> 
> https://www.geeksforgeeks.org/constructor-delegation-c/

aaa, ok :)

File: shapes/Shape.hpp [hunk 8]
```diff
@@ -8,8 +8,9 @@ enum class Color : unsigned char {
 
 class Shape
 {
-    Color color_ = Color::Red;
 public:
+
+    Color color_ = Color::Red;
```
Comment (ziobron) [inline shapes/Shape.hpp]: it should stay private

File: shapes/Rectangle.cpp [hunk 9]
```diff
@@ -4,6 +4,15 @@
 Rectangle::Rectangle(double x, double y, Color c) : x_(x), y_(y),  Shape(c)
 {}
 
+Rectangle::Rectangle() : x_(1), y_(1)
+{}
+
+
+Rectangle::Rectangle(Color c) : Rectangle()
```
Comment (ziobron) [inline shapes/Rectangle.cpp]: we want to have such 2 constructors:
`Rectangle(double x, double y, Color c);  // already implemented`
`Rectangle(double x, double y);  // not implemented`
The latter should call the first one on it's initializer list.

File: shapes/Rectangle.cpp [hunk 10]
```diff
@@ -4,6 +4,15 @@
...
+{
+    color_ = c;
```
Comment (ziobron) [inline shapes/Rectangle.cpp]: you don't need to set it here. Use the constructor `Rectangle(double x, double y, Color c)`. Body will be empty.

**Review by ziobron — CHANGES_REQUESTED** [review]

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
15 XP za Nowoczesny C++ SFINAE (4447)
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ delegating constructors (4449)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4447, 4448, 4449, 4450, 4451, 4452] }

**Pull request closed without merging**